File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/borders/radius/BorderRadiusConfig.swift
//
//  BorderRadiusConfig.swift
//  StyleEngine/borders/radius — Phase 5.
//
//  Value struct for CSS `border-{top,right,bottom,left}-{left,right}-radius`
//  physical corners and the four logical equivalents
//  (`border-{start,end}-{start,end}-radius`). Each corner carries an
//  independent x / y radius so elliptical corners ("20px 10px") survive
//  round-trip — iOS's `RoundedRectangle(cornerRadius:)` only supports
//  square corners so we ship a custom Shape alongside (see
//  `BorderRadiusShape.swift`).
//
//  Mirrors Android's `radius/BorderRadiusConfig.kt`; the logical-to-
//  physical mapping assumes LTR writing mode (matches what the fixtures
//  exercise).
//

// CoreGraphics for `CGFloat`, SwiftUI transitively for Shape.
import SwiftUI

// Per-corner radii. Either side of the pair can be zero independently
// which is how CSS communicates a sharp corner on one axis and a curve
// on the other (`border-radius: 20px 0 0 0 / 10px 0 0 0`).
struct BorderRadiusCorner: Equatable {
    // Horizontal radius in points.
    var x: CGFloat = 0
    // Vertical radius in points.
    var y: CGFloat = 0
    // Percent radii (0–100). CSS Backgrounds 3 §5.1: "Percentages: refer
    // to corresponding dimension of the border box" — the horizontal
    // radius resolves against the box WIDTH, the vertical radius against
    // the box HEIGHT. We can't know the box size at extract time, so a
    // percent axis is carried symbolically and resolved by
    // `resolved(in:)` at draw time. `border-radius: 50%` on a 200×80 box
    // must produce a 100×40 elliptical corner (web behaviour) — NOT the
    // min(w,h)/2 capsule the old px-only placeholder produced.
    var xPercent: CGFloat? = nil
    var yPercent: CGFloat? = nil
    // True when every axis is zero/absent — the Shape drops the arc
    // altogether when both are zero so we get a crisp right angle.
    var isSquare: Bool { x <= 0 && y <= 0 && xPercent == nil && yPercent == nil }
    // Convenience for the common "circular corner" case.
    init(uniform r: CGFloat) { self.x = r; self.y = r }
    // Elliptical init — kept distinct from `uniform` for readability.
    init(x: CGFloat, y: CGFloat) { self.x = x; self.y = y }
    // Percent-aware init — either axis may independently be a percent.
    init(x: CGFloat, y: CGFloat, xPercent: CGFloat?, yPercent: CGFloat?) {
        self.x = x; self.y = y
        self.xPercent = xPercent; self.yPercent = yPercent
    }
    // Zero init used by the default-struct synthesis.
    init() {}

    /// Resolve this corner against a concrete border-box size. Percent
    /// axes multiply width (x) / height (y) per CSS Backgrounds 3 §5.1;
    /// absolute axes pass through unchanged.
    func resolved(in size: CGSize) -> BorderRadiusCorner {
        BorderRadiusCorner(
            x: xPercent.map { size.width * $0 / 100 } ?? x,
            y: yPercent.map { size.height * $0 / 100 } ?? y
        )
    }
}

// Four-corner bundle. The extractor maps logical corners into these
// buckets using the LTR assumption:
//   border-start-start-radius  → topLeft
//   border-start-end-radius    → topRight
//   border-end-end-radius      → bottomRight
//   border-end-start-radius    → bottomLeft
struct BorderRadiusConfig: Equatable {
    // North-west corner.
    var topLeft: BorderRadiusCorner = BorderRadiusCorner()
    // North-east corner.
    var topRight: BorderRadiusCorner = BorderRadiusCorner()
    // South-east corner.
    var bottomRight: BorderRadiusCorner = BorderRadiusCorner()
    // South-west corner.
    var bottomLeft: BorderRadiusCorner = BorderRadiusCorner()

    // Derived: true when at least one corner carries a non-zero radius.
    // Lets appliers skip the Shape allocation for the common square case.
    var hasAny: Bool {
        !topLeft.isSquare || !topRight.isSquare
            || !bottomRight.isSquare || !bottomLeft.isSquare
    }

    // Largest radius on the box — used by the outline applier for the
    // outset-follow-radius case. Percent axes are not resolvable here
    // (no box size in scope) and contribute their raw px component (0);
    // the outline path tolerates the underestimate.
    var maxRadius: CGFloat {
        max(topLeft.x, topLeft.y, topRight.x, topRight.y,
            bottomRight.x, bottomRight.y, bottomLeft.x, bottomLeft.y)
    }
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/borders/radius/BorderRadiusConfig.kt
package com.styleconverter.runtime.borders.radius

import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * Configuration for border radius on all four corners.
 *
 * CSS border-*-radius accepts *two* values per corner — an x-axis radius
 * followed by a y-axis radius — producing elliptical corners. The IR
 * encodes each corner as {"horizontal": IRLength, "vertical": IRLength}
 * (parsed in src/main/kotlin/app/parsing/css/properties/longhands/borders/radius/).
 *
 * We store each corner as an (x, y) pair of Dp. When x == y the corner is
 * circular (the common case), and [isCircular] is true so the applier can
 * take the fast path using [androidx.compose.foundation.shape.RoundedCornerShape].
 * When x != y the applier uses a custom Shape that draws elliptical corners.
 *
 * Start/end naming aligns with Compose's RTL-aware corner naming:
 *  - topStart = top-left in LTR, top-right in RTL
 *  - topEnd   = top-right in LTR, top-left in RTL
 *  - bottomEnd   = bottom-right in LTR, bottom-left in RTL
 *  - bottomStart = bottom-left in LTR, bottom-right in RTL
 */
data class BorderRadiusConfig(
    // Each corner is a (horizontal, vertical) radius pair. x==y means a
    // circular corner; x!=y means elliptical. Defaults are 0 (square).
    val topStart: Pair<Dp, Dp> = 0.dp to 0.dp,
    val topEnd: Pair<Dp, Dp> = 0.dp to 0.dp,
    val bottomEnd: Pair<Dp, Dp> = 0.dp to 0.dp,
    val bottomStart: Pair<Dp, Dp> = 0.dp to 0.dp,
    // PERCENTAGE radii, one nullable fraction (0..1) per axis per corner.
    // css-backgrounds-3 §4.4: a percentage horizontal radius refers to the
    // WIDTH of the border box and a vertical one to its HEIGHT — both only
    // known at paint time, so percentages ride along as fractions and the
    // applier's Shape resolves them against the laid-out size. When an
    // axis fraction is non-null it WINS over the Dp in the pair above
    // (the extractor writes a 0.dp placeholder there). This is what makes
    // `border-radius: 50%` on an unsized box render as a full ellipse
    // (Borders_Decorated: web ellipse vs Android's old plain rectangle,
    // SSIM 0.7746 — percent radii used to be dropped entirely when the IR
    // carried no Width/Height to resolve against).
    val topStartFraction: Pair<Float?, Float?> = null to null,
    val topEndFraction: Pair<Float?, Float?> = null to null,
    val bottomEndFraction: Pair<Float?, Float?> = null to null,
    val bottomStartFraction: Pair<Float?, Float?> = null to null
) {
    /** True when any corner carries a paint-time percentage axis. */
    val hasFraction: Boolean
        get() = listOf(topStartFraction, topEndFraction, bottomEndFraction, bottomStartFraction)
            .any { it.first != null || it.second != null }

    /**
     * True if any corner has a non-zero radius on either axis (fixed Dp or
     * percentage). The applier short-circuits when this is false so we
     * don't allocate a Shape for square-cornered elements.
     */
    val hasRadius: Boolean
        get() = topStart.first > 0.dp || topStart.second > 0.dp ||
                topEnd.first > 0.dp || topEnd.second > 0.dp ||
                bottomEnd.first > 0.dp || bottomEnd.second > 0.dp ||
                bottomStart.first > 0.dp || bottomStart.second > 0.dp ||
                hasFraction

    /**
     * True when every corner is circular (x == y on every corner). When
     * true (and no percentage axes are present) the applier can use
     * RoundedCornerShape instead of a custom Shape.
     */
    val isCircular: Boolean
        get() = topStart.first == topStart.second &&
                topEnd.first == topEnd.second &&
                bottomEnd.first == bottomEnd.second &&
                bottomStart.first == bottomStart.second

    companion object {
        val NONE = BorderRadiusConfig()
    }
}
